src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,506 @@
+    this.executor.setContinueExistingPeriodicTasksAfterShutdownPolicy(false);
+    this.executor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+    this.executor.setRemoveOnCancelPolicy(true);
+
+    //Create counter used to restrict the number of messages we fetch every second.
+    this.bucket = new AtomicInteger();
+
+    //Configure properties
+    this.project = config.getProject();
+    this.topic = TopicName.format(config.getProject(), config.getTopic());
+    this.subscription = ProjectSubscriptionName.format(config.getProject(), config.getSubscription());
+
+    try {
+      this.credentials = config.getServiceAccount() == null ?
+        null : GCPUtils.loadServiceAccountCredentials(config.getServiceAccount(),
+                                                      config.isServiceAccountFilePath());
+    } catch (IOException e) {
+      LOG.error("Unable to get credentials.");
+      stop("Unable to get credentials for receiver.", e);
+    }
+
+    //Create subscription if the topic is specified.
+    if (topic != null) {
+      createSubscription();
+    }
+
+    //Schedule tasks to set the message rate (Token Bucket algorithm) and start the receiver worker.
+    this.executor.scheduleAtFixedRate(this::updateMessageRateAndFillBucket, 0, 1, TimeUnit.SECONDS);
+    this.executor.submit(this::scheduleFetch);
+
+    LOG.info("Receiver started execution");
+  }
+
+  @Override
+  public void onStop() {
+    //Shutdown thread pool executor
+    if (executor != null && !executor.isShutdown()) {
+      executor.shutdown();
+      try {
+        executor.awaitTermination(30, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        LOG.error("InterruptedException while waiting for executor to shutdown.");
+      }
+    }
+
+    //Clean up subscriber stub used by the Google Cloud client.
+    if (subscriber != null && !subscriber.isShutdown()) {
+      subscriber.shutdown();
+      try {
+        subscriber.awaitTermination(30, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        LOG.error("InterruptedException while waiting for subscriber to shutdown.");
+      }
+    }
+
+    LOG.info("Receiver completed execution");
+  }
+
+  /**
+   * Create a new subscription (if needed) for the supplied topic.
+   *
+   * @throws IllegalArgumentException when the specified Topic does not exists
+   * @throws RuntimeException         when the SubscriptionAdminClient cannot be created.
+   * @throws ApiException             when a non-retryable exception is thrown by the Pub/Sub client
+   */
+  protected void createSubscription() {
+    int backoff = backoffConfig.getInitialBackoffMs();
+    int attempts = 5;
+
+    ApiException lastApiException = null;
+
+    while (!isStopped() && attempts-- > 0) {
+
+      try (SubscriptionAdminClient subscriptionAdminClient = getSubscriptionAdminClient()) {
+
+        int ackDeadline = 10; //10 seconds before resending the message.
```

Comment:
still have this question

Reply:
Good point. I did not notice this in my testing, but we can always set this timeout to 60 seconds.

I can add a config option for this if you think this would be useful.